### DISCUSSION TITLE
Web Checkout URL Without Login

### DISCUSSION THREAD

**Discussion by archimedes287** [Shopify/storefront-api-feedback#105] [Ideas & Requests]
### FEATURE REQUEST

Thanks in advance for you help!

The Shopify GraphQL Storefront API is awesome. Unfortunately, it has a huge user experience limitation that is quite PAINFUL to run into after developing a custom storefront for Shopify.

```
pain·ful
/ˈpānfəl/
adjective
(of a part of the body) affected with pain.
```
### LIMITATION

If you complete a checkout for a logged in customer, then the customer is prompted to log in again.

You can't use the `X-Shopify-Customer-Access-Token` header to preserve authentication when the customer is associated to the checkout.

Multipass login requires additional server side setup and costs $2,000 a month.

Your custom storefront is not truly headless because the Shopify login pages are required for checkout.

https://shopify.dev/api/examples/checkout#complete-a-checkout-for-a-logged-in-customer

### SECURITY

It looks like the security concern is stated here: 'https://github.com/Shopify/js-buy-sdk/issues/561#issuecomment-420756486'

The `customerAccessToken` can also be _"shared or stored [and gives] direct access to customer details"_ so I'm a little confused about why the unauthenticated `webUrl` is any less secure than the `customerAccessToken`. Could you just limit the number of times the `webUrl` is able to be used for a checkout request to tighten up the security issues? Similarly, could you limit the time that the `webUrl` is valid for?

### POSSIBLE SOLUTION

If a `webUrl` is less secure for some reason, then would it be secure to put the `customerAccessToken` or some other token in a cookie that the web checkout form, living on a subdomain, could access to log the customer in?

```javascript

// CLIENT (Web browser)
// set a cookie the subdomain can access with the customerAccessToken
// redirect to web checkout

var custom_storefront_domain = "https://custom-domain.com/";
var shopify_checkout_domain = "https://checkout.custom-domain.com/";

storefrontAPI.getCartWithBuyerIdentity(cart) => {
    document.cookie = "customerAccessToken=tEv2Ew6gbBVoKSALR8Xgp6lWrgYSlIi2;path=/;domain=custom-domain.com";
    window.location.href = cart.checkoutUrl;
});

```
and
```javascript
// Storefront API Server

app.get('/fancyUrl', function(req, res){
    var customerAccessToken  = req.cookies.customerAccessToken;
    if (tokenCustomerMatchesUrlCustomer) {
        skipLoginAndGoToCheckout();
    } else {
        redirectToSomeUrl();
    }
});

```

**Comment by littleleoman** [comment]
Currently facing the same issue which makes it impossible to have a complete storefront if customer login is required prior to checkout. I don't have much to add because you've pretty much covered everything so just adding this here in hopes that this request gains traction.

On my end, I have also tried making use of the **_cartBuyerIdentityUpdate_** mutation after customer login on my hydrogen storefront, but that ultimately had no effect in the checkout flow (once my user clicks checkout on the hydrogen app, it redirects them to the shopify hosted site (example.myshopify.com) login screen).

**Reply by qsales2022** [reply]
> Currently facing the same issue which makes it impossible to have a complete storefront if customer login is required prior to checkout. I don't have much to add because you've pretty much covered everything so just adding this here in hopes that this request gains traction.
> 
> On my end, I have also tried making use of the **_cartBuyerIdentityUpdate_** mutation after customer login on my hydrogen storefront, but that ultimately had no effect in the checkout flow (once my user clicks checkout on the hydrogen app, it redirects them to the shopify hosted site (example.myshopify.com) login screen).

bro did you find any solution?

**Comment by zwergius** [comment]
I recently refactored one of my stores from checkout to cart using storefront api. When using the checkout I managed to get the email registered and showing up on the checkout page. This I haven't accomplished on the new Cart api. I wonder why that mutation to add customerAccessToken is even there, it seems it has no effect at all.

**Reply by electricenjindevops** [reply]
Completely agree... Baffled. I have their access token, I don't understand why I can't log the user ON PAGE in to the store when I literally have their access token created by email/pw. Seems like a major oversight from Shopify to force users to use only the normal signup or login - god forbid you have any sort of flow that needs this inside of another flow.

**Comment by vanska** [comment]
I'm also **extremely** interested to see a nice solution for this issue. Even the Multipass route seems unnecessarily complicated and does indeed require the Plus subscription which, let's be frank is not an option for many store owners.

There's also a [thread open on Shopify community](https://community.shopify.com/c/storefront-api-and-sdks/shopify-headless-cart-implementation-requires-users-to-login/td-p/1428870) on this exact same topic that hasn't gained any traction from the Shopify side.

Is this something the storefront product team is thinking about/planning to fix and when might we see a resolution @vixdug? It's little things like this that make it harder to push headless Shopify to customers that are actively asking for it 🤷‍♂️.

Shouldn't `cartBuyerIdentityUpdate` do exactly what we want?

**Comment by johncraigcole** [comment]
Hey Everyone 👋 Just wanted to mention having an authenticated buyer on a custom storefront be persisted over to the Shopify web checkout is something we've heard, and is being worked on. I can't share any timeline as of yet, but we know this is really important to the community and we want to get this right for all merchant types.

The `cartBuyerIdentityUpdate` mutation currently will only bring the buyers email into the Shopify web checkout (not logged in), and it's primary role is about establishing customer context for the active cart. Customer context from this mutation will have an impact on pricing (the market associated with the customer), and delivery group options.

I know it's not an authenticated buyer that everyone here is after, but I wanted to add that on navigation to the Shopify web checkout, checkout form fields can be appended as [query string parameters](https://shopify.dev/docs/apps/checkout/cart-permalinks/cart-permalinks#supported-checkout-parameters) to the `checkoutUrl` of the cart object. This can help reduce data entry for buyers not using an accelerated payment option like Shop Pay.

**Reply by alevals** [reply]
hello @johncraigcole do you have any updates on the progress of this ?

**Reply by johncraigcole** [reply]
Hey @alevals 👋 , work continues and this remains a high priority to get solved.

**Reply by slavanossar** [reply]
The link in your comment is broken, [this](https://shopify.dev/docs/apps/checkout/cart-permalinks/cart-permalinks#supported-checkout-parameters) one is working.

Not sure if I'm missing something obvious, but appending a query string with those values to the `checkoutUrl` that is returned by Storefront GraphQL API when querying the `cart` object has no effect on pre-filling the fields.

The docs you link to specify a `https://{shop}.myshopify.com/cart/...` URL format, while the `checkoutUrl` on the `cart` object looks like `https://{shop}.myshopify.com/checkouts/...`.

**Reply by johncraigcole** [reply]
Hey @slavanossar thanks for reporting the bad link. I've tested this again and appending `?checkout[shipping_address][first_name]=john` to the `checkoutUrl` of the Storefront API cart, and then visting that url does set the first name field of the shipping address to `john` in my test.

If you've upgraded to [Shopifys checkout extensibility platform](https://help.shopify.com/en/manual/checkout-settings/checkout-extensibility), information provided to the `cartBuyerIdentityUpdate` [mutation](https://shopify.dev/docs/api/storefront/2022-10/mutations/cartBuyerIdentityUpdate) like `email` or `deliveryAddressPreferences` will also pre-fill the values at checkout, but still does not log the customer in at checkout. The logged in piece is still important I just can't share a timeline on when that piece will be available.

**Reply by electricenjindevops** [reply]
Can you please update us ASAP. It is very odd that there is no way to log a customer in. In our case, we're not even using headless... We just want to implement a custom login on the hosted Shopify site and without Multipass this seems impossible. We have sign up built in to our flow for something else, not using normal customer sign up, and we want to log them in automatically because they've just entered their email and password. We're having a hard time understanding why this is not possible without basically creating a weird workaround for the regular login form.

**Reply by johncraigcole** [reply]
@electricenjindevops if Shopify is not managing authentication (you're using a 3rd party),  in the future the new customer accounts and public api will support authentication via a 3rd party. I don't have a timeline to share on when this will be available. Today, the only way to implement your own auth, and have the customer be logged in, is only possible for a Plus merchant with multipass.

**Reply by tobbbe** [reply]
@johncraigcole would it be possible to add phone number to [checkout parameters](https://shopify.dev/docs/apps/checkout/cart-permalinks/cart-permalinks#supported-checkout-parameters)? 😃

**Reply by johncraigcole** [reply]
Hey @tobbbe 👋 . While not documented, phone `?checkout[shipping_address][phone]=12345678` does work. Again, if your store is already on the checkout extensibility platform I would stick to using the `deliveryAddressPreferences` on the `cartBuyerIdentityUpdate` mutation to communicate any and all address fields.

**Reply by tobbbe** [reply]
Oh okey! Gonna try that then. Thanks!!

Unfortunately this project isn't on a Plus plan so I can't use checkout extensibility platform.

**Reply by johncraigcole** [reply]
While someone not on a Plus can't use something like a checkout ui extension, their storefront is likely still on the latest version of checkout which should support the `deliveryAddressPreferences`. If you have the time I would take a look. Reminder that the inputs to `deliveryAddressPreferences` provided via `cartBuyerIdentityUpdate` must provide all the required fields.

**Reply by tobbbe** [reply]
Ohh! Awesome! Trying that tomorrow.
Thanks a lot for the info that all fields must be set. I tried to set country today with that mutation and it didn't work. Probably because I didn't set all then.

**Reply by timfuhrmann** [reply]
Hey @johncraigcole, quick follow up question, is this planned to remain exclusive for new customer accounts? While I would love to not have to use multipass, we can't get lost of classic accounts yet.

**Comment by hanayashiki** [comment]
Why not try this API?
https://shopify.dev/docs/api/storefront/2023-07/mutations/checkoutCustomerAssociateV2

**Reply by hanayashiki** [reply]
tried. no luck.
It seems having an auth cookie at shopify domain is the only chance

**Comment by arobbins** [comment]
@johncraigcole Would love to have an update on this when you have time.

**Comment by johncraigcole** [comment]
Thanks for the ping @arobbins . At summer editions, we actually introduced our [new customer account api](https://shopify.dev/docs/custom-storefronts/building-with-the-customer-account-api/?itcat=summer-23-editions-page&itterm=CustomerAccountAPI) that will support auth into checkout. 

We're early days and the Customer account API is still in unstable, but we're targeting it's first stable release for `2023-10`. Initially, we only released an example implementation that works with the [Hydrogen channel](https://shopify.dev/docs/custom-storefronts/building-with-the-customer-account-api/getting-started), and uses our new customer accounts for password-less auth. More examples will be coming in the future, and the plan is to also support 3rd party auth providers but I don't have a timeline on that.

For headless merchants not using the Hydrogen channel, you'll need to use our Headless channel instead of configuring a custom app that has Storefront API access. The headless and hydrogen channels provide access to the the necessary customer account type settings like Callback uri's, origin, logout url etc. In addition to using the headless channel, there are some other important requirements.

1. The customer account settings in the admin (`settings/customer_accounts`) have a version of `new customer accounts`. 
2. Cart needs to be built with the Storefront API Cart object. Once logged in on the headless site via the customer account api, you'll append `logged_in=true` as a query string parameter to the `checkoutUrl` of the Storefront API cart. If the buyer is authenticated, it will carry into checkout. `logged_in=true` does not work with the Storefront API checkout object, and only works if the auth happens with the customer accounts api.
3. The store is on the latest checkout often referred to as the checkout extensibility platform. For non Plus merchants, this is likely already to be the case. For Plus merchants, there might still be a good percentage still using checkout.liquid. If you're a plus merchant you can validate you're on the extensibility platform by reviewing the Checkout settings in the admin to see if you have a published checkout profile found under the Checkout customization card. Before publishing a checkout profile, please ensure that all checkout customizations formerly in checkout.liquid have been ported over to the checkout extensibility platform and tested.

**Reply by arobbins** [reply]
Thank you for the thoughtful update. This is really exciting.

**Reply by tbaba19** [reply]
hi
how are you doing

**Reply by iamkyle** [reply]
Hello, John, 
It appears as though the logged_in=true feature no longer works with the new customer accounts api. Is there any newer documentation on getting this to work as expected, or any way to associate the buyerIdentity with a customer authenticated through the customer accounts api?

**Reply by johncraigcole** [reply]
Hey @iamkyle I just ran some tests and all looks good to me. Would need a little more context on where things are breaking down. Did the authentication on shopify.com succeed and you were to able generate an access token to be used with the new customer accounts api and it's just the `logged_in=true` being appended to the Storefront API cart checkout url that isn't working? 

Could mean the store isn't on the checkout extensibility platform or new customer accounts are not turned on in the customer settings section of the admin.

**Reply by arobbins** [reply]
Hey John, another question for you if you have time.

In your above comment earlier you mentioned this:

`For headless merchants not using the Hydrogen channel, you'll need to use our Headless channel instead of configuring a custom app that has Storefront API access.`

I have a custom app that generates a Storefront Access Token, which I then use to manage the user's cart session, etc. 

If I understand correctly... since I'm not using Hydrogen this feature is currently unavailable to me?

Is there any timeline estimate of when we can expect availability for custom sales channels?

**Reply by iamkyle** [reply]
@johncraigcole apologies for the lack of context. 

Yes, the authentication succeeds and access code is generated. logged_in=true is also being appended to the storefront cart checkout url. This specific problem is occurring for B2B customers. When they arrive at the cart, it doesn't look like the buyerIdentity is set and they do not have access to their business information (i.e. payment terms, options, etc...)

The other settings are also confirmed - checkout extensibility and new customer accounts.

**Reply by johncraigcole** [reply]
Hey @iamkyle if you haven't already please contact support so we could start the process of looking into this further. Out of curiosity, if you were to go back to the custom storefront and follow the Storefront API cart `checkoutUrl` again a couple of times does it eventually load the B2B context you're expecting? 

The customer account api itself just recently went into release candidate for `2024-01` from your comment was this working fine even though we were in `unstable` of the customer account api until recently?

**Reply by johncraigcole** [reply]
Hey @arobbins correct. The feature as it stands right now is limited to the Hydrogen and Headless channels where the additional configuration needed for the customer account api can be specified. I'll see if I can find if there are any plans to add support for this outside the Hydrogen and Headless channels.

**Reply by iamkyle** [reply]
@johncraigcole Yes, I have contacted support regarding this issue. Just trying all the avenues i can to get this resolved. I'm sure you know how frustrating the support channels can be when you're both the store owner and developer... And yes, it does eventually flip over to the business context. I haven't been able to find a consistently reproducible way to do this, though. Sometimes it works when you click the "Log out" link in the checkout itself. Sometimes it works when you navigate back to cart, add something to the cart, and then continue to the checkout link. The inconsistency is the most challenging part as we can't just say "do this and it will work". I was happy to see that it went into release candidate status. We were in the early stages of development of our new storefront when the new api was announced and i didn't want to build something on an outdated foundation to start. figured we could live on shaky ground for a couple of months if the fundamentals weren't going to change. i am not relying on much of the api for anything other than authentication, which seemed to be pretty fundamental to the release and was fairly clearly documented. would be happy to jump into slack and screen share to show you what exactly is happening.

**Reply by iamkyle** [reply]
@johncraigcole for what it's worth, it seems like it works when the cart buyerIdentity customer is set. When I do the back/forth magic it looks like the buyer identity eventually gets set on checkout and then it works.

**Reply by johncraigcole** [reply]
Hey @iamkyle thanks for the context. I was working with a colleague yesterday in support on a similar issue. What I believe the issue is today is that we don't have a way to associate B2B buyer context with the Storefront API cart. So while carrying the authenticated buyer into checkout works, the underlying cart has no B2B Context. This seems to get corrected in going back and forth but that isn't a usable flow. It's definitely on the roadmap to add B2B context support to the Storefront API cart but I don't have a timeline to share on when that will be available.

**Reply by iamkyle** [reply]
@johncraigcole Ack... it seemed as though B2B was one of the primary drivers for the new customer accounts api. And one of the reasons we pushed ahead with Shopify Plus. Big blow. 

This feature was also previously working as expected. Up until just a couple weeks ago we could consistently go through checkout. 

Is there anything I can do to force the cart context? it seems as though i just need to update the cart buyerIdentity (every time that's set in the cart, it seems to work as expected). Can I exchange a customer account session for a storefront session and then set the buyerIdentity?

**Reply by johncraigcole** [reply]
Hey @iamkyle have you actually been using the Storefront API `cartBuyerIdentityUpdate` mutation? When you say it was working up until a few weeks ago were you on new customer accounts, and the checkout extensibility platform at that time? Assuming you were on new customer accounts and checkout extensibility at the time it was working, you're indicating that when following the Storefront API cart `checkoutUrl` and appending `logged_in=true` that the B2B context was working in checkout on initial arrival to checkout.

Given you've already contacted support, it would probably make more sense to move this discussion there where you've been authenticated on the store and we can discuss specific store context more freely than we can here. Do you have the ticket number?

**Reply by iamkyle** [reply]
No, the cartBuyerIdentityUpdate does not work with session tokens created in the new customer account api. I'm saying I have not been able to get the buyerIdentity to update for the cart. Up until it stopped working, all i had to do was follow the checkoutUrl with logged_in=true. 

Right now the most reliable way to trigger the B2B context is actually to click on "Log out" in the checkout. Every time i do that it doesn't log out, but instead sets the correct context. 

I really appreciate you responding here. it's very helpful to actual talk with someone. I will get the support ticket number and provide that shortly.

**Reply by iamkyle** [reply]
@johncraigcole Ticket #43978938

**Reply by johncraigcole** [reply]
I just wanted to close this out with an update. Currently, `12/18/2023` we don't yet have support for contextualized B2B context on the Storefront API cart. This is being worked on, but I don't have a timeline I can share on when it will be available. 

The problem experienced by Kyle in using new customer accounts, where the B2B context of the logged in customer was not being loaded at checkout has been resolved.

**Reply by miksony** [reply]
@johncraigcole  Hello, I have a problem with authenticating the user on a checkout. The user is sometimes logged in and sometimes not. 

I am using Storefront API with a headless channel and a custom store written in React. Also, I am using the `Shopify/hydrogen` package. I have a `CartProvider` wrapper that uses `customerAccessToken` that is retrieved with `customerAccessTokenCreate` mutation. 

Problem with checkout: 
1. Login on the headless store and put items in the cart
2. Redirect to checkout with `logged_in=true` appended to the checkout URL
3. On a first visit user **is logged in** in a Shopify checkout, when I go back to the store and go to the checkout again, the user **is not logged in** only the email is pre-filled. If I change the user, nothing happens. I see that `cartBuyerIdentityUpdate` is called regularly, so the new user is always provided. 
4. If I clear all site data on checkout next time I visit it user **is logged in**. 

What can be the issue?

Currently, we have Shopify Plus and using a multipass but that is not more option for us, so we are migrating to a non-Shopify Plus account and need to have the user logged in into the checkout.

**Reply by mzielu** [reply]
I have the same issue as @miksony. 

1. I have a cart created with `cartCreate` mutation.
2. Later I am updating **buyerIdentity** field using `cartBuyerIdentityUpdate` mutation.
3. Later I am using the simplest `cart` query to fetch the **checkoutUrl** which is included inside the response
4. Opening the given URL at first try works like a charm and I have data prefilled in the checkout.
5. When I want to open the checkout URL again data is not prefilled at all. I see that **buyerIdentity** becomes null again in that situation so I perform `cartBuyerIdentityUpdate` mutation again but the result is the same. Checkout data is not prefilled at all.

@johncraigcole can you help resolve the issue?

**Reply by vixdug** [reply]
Hi @mzielu - looking into this and I'm not able to replicate. Can you send me a request-id (from the response headers) of your most recent attempt where buyerIdentity was not populated?

**Reply by mzielu** [reply]
@vixdug I have recorded a video for you to show what is happening on frontend (request ids in the brackets)

https://github.com/Shopify/storefront-api-feedback/assets/15313583/92431566-7df3-4070-b22c-6aa32726d31b

1. first I perform `cart` query to get **checkoutUrl**, **buyerIdentity** is not null (f2587710-d3b5-4b9d-b69f-6777d63d861b-1715793435)
2. I open the checkout URL and I see the prefilled data
3. I go back to shopping cart screen and refresh the cart using `cart` query, **buyerIdentity** is still not null (c8e4ea94-13ae-4b89-bb43-fc0f4d6ee101-1715793451)
4. I open the checkout URL and I see that data is not prefilled
5. I go back to shopping cart screen and refresh the cart using `cart` query (ca91d8b3-befe-437a-85ca-45085e99b12d-1715793461), **buyerIdentity** is NULL now so I update it with `cartBuyerIdentityUpdate` mutation (
46a76097-35cc-4aa6-83f1-e717ba8806f1-1715793462)
6. I open the checkout URL and I see that data is not prefilled

**Reply by mzielu** [reply]
It seems that checkoutUrl after the first execution "is broken" and when the user tries to execute it later **buyerIdentity** data is ignored. I have tested this on mobile and web and in both cases the result is the same.

**Reply by kaisegni** [reply]
+1

**Reply by mzielu** [reply]
@vixdug any updates on the above issue? I see that **checkoutURL** is being redirected from:
`https://{shop}/cart/c/............... `to: `https://{shop}/checkouts/cn/.........`

Can it be a potential source of the problem?

**Reply by miksony** [reply]
@vixdug @johncraigcole  I have a similar problem explained in the comments above. Any update on it?

**Comment by electricenjindevops** [comment]
Hey, I think you are misunderstanding. A lot of us are trying to use
Shopify's customer accounts to log in to Shopify, on Shopify... Can you
explain how graphql on a hosted Shopify store in a liquid theme can be used
to log a customer in? There is email and password graphql for the customer
access token, but there seems to be no way to actually log the customer in
with it.

I am talking about normal Shopify accounts here. Not third party auth, and
not Multipass.

**Reply by johncraigcole** [reply]
The Storefront API can not be used to a log a customer in to the online store. It can only be used to generate an acces token from which you can then query for information about the customer, and make changes to the customer such as adding additonal addresses.

I'm curious about the use case of not wanting to use the default login Shopify provides with online store.

**Reply by electricenjindevops** [reply]
They sign up in a custom form, we pass that to our backend, we create the Shopify account for them (and other accounts / data saves), and then we return success to frontend. In this case, we got it working - because the user just signed up, they have their plaintext password/email in the form already (and we never save their password in plaintext) so we can trigger auto-login from Javascript using the form data. We got it all working, but wish this had been more well-documented. The use case is a custom signup leading to an automatic login without a page refresh / redirect / etc. It works currently for us.

**Comment by electricenjindevops** [comment]
Figured it out, finally. Why this is not more well documented is beyond me.
You just send a post to /account/login with correct form data and
form_type=customer_login and it logs them right in.

The use case is we ask user to create an account, we send to our own API.
In our own api, we create a Shopify account, a Salesforce profile, a Venly
address, and more. Now, finally, we can silently log them in to their brand
new Shopify account automatically before they even finish the next page of
our form wizard.

I cannot for the life of me figure out why on-site Shopify login with JS is
not more publicly documented. Forcing stores to use official login and sign
up pages would eliminate so many use cases. Yes, we use Shopify, but it's
not the main account for our customers and its part of a bigger user flow
so it was never an option to redirect them to the normal sign up.

**Reply by hanayashiki** [reply]
That requires you to save the user's password in plain text, isn't it?

**Reply by electricenjindevops** [reply]
Yes to plaintext data, no to the saving. We have a signup form inside of
another process, and a login form they can use, so this is just a normal
POST request. Their request comes through and we create a Shopify account
from our API using their new credentials (or if they log in, they don't
send us a password at all & our API just handles non-account related data),
and then their formdata is used for the login on Shopify's side.

**Reply by julianCast** [reply]
I'm facing the same issues with not finding a way to log in a customer via API. One question about your approach:
From where are you making the post request to /account/login ? A custom app? From within the storefront? Or from a external site? Because I'm getting CORS errors from client (I understand) and if it's coming from a server the response I get from the post call is the website text. How do you redirect to it? Thanks!

**Reply by arobbins** [reply]
Is this still working for you? @electricenjindevops

**Comment by electricenjindevops** [comment]
On ours, this is what we did:
- Registration form ON shopify
- Send details to our backend API, via a shopify proxy app
- Create accounts on Shopify, Auth0, Salesforce, Venly
- Send success message back to frontend
- Because the user filled out the form w/ their email & password ON
shopify, we just take the form values and submit to the login route ON
shopify.

**Comment by Judimax** [comment]
Greetings I am not sure if this is solved but essentially from what I understand you create the customer access token from their email and password
```    
mutation customerAccessTokenCreate($email:String!,$password:String!) {
      customerAccessTokenCreate(input: {email: $email, password: $password}) {
        customerAccessToken {
          accessToken
        }
        customerUserErrors {
          message
        }
      }
    }
``` 

then you create the cart and open the checkoutUrl 

```
    mutation cartCreate($input: CartInput!) {
      cartCreate(input: $input) {
        cart {
          id
          checkoutUrl
        }
        userErrors {
          field
          message
        }
      }
    }
```

with these variables
```
    variables = {
      "input":{
        "buyerIdentity": {
          "email":email,
          "customerAccessToken": customerAccessTokenFromEarlier
        },
        "lines":[
          {
            "merchandiseId":merchandiseId,
            "quantity": quantity,
          } ...
        ]
      }
    }
``` 
so this will even be better becuase instead of email it connect to the correct shopify account 

My question is (unrelated)
once they complete the purchase how would they redicted away from the order status page to another page 
Is there an undocumented query parameter that allows for this?

**Reply by hanayashiki** [reply]
Doesn't work for me.
Using a `customerAccessToken` that works for GraphQL Storefront, and I have "logged-in required for checkout" set, and make sure the user hasn't logged into the Shopify side

```gql
mutation CreateCart {
  cartCreate(
    input: {buyerIdentity: {email: "wangchenyu2017@gmail.com", customerAccessToken: ",,,"}, lines: [{quantity: 1, merchandiseId: "gid://shopify/ProductVariant/..." }]}
  ) {
    cart {
      checkoutUrl
    }
  }
}
```

Then I copy the checkout url to the browser. Still, the user is redirected to the login page in the liquid store instead of going to checkout with an identify associated with `customerAccessToken`.

Update: I was using 2023-10 but 2024-4/2024-7 also doesn't work for me.

**Reply by Judimax** [reply]
im not familiar with shopify liquid, there is also the cart url in the docs you will have to double check.
Also depending on the app type (private public custom, headless) will you have certain issues 
For storefront im using private app

**Reply by arobbins** [reply]
I believe this only works for the Hydrogen sales channel.

**Comment by arobbins** [comment]
@johncraigcole 

Since it's been almost a year, I was wondering if you could provide an update on when this should start working with the Storefront API—without Hydrogen?

As a reminder of my setup:
- I have a custom Shopify app that uses the Storefront API to manage Carts for my users
- My users come from a custom WordPress plugin I've made called ShopWP. (https://wpshop.io). The frontend lives on WordPress.
- I need to "log in" the user from the WordPress site
- When the user is ready to checkout, they need to remain logged in while on the Shopify checkout page
- I'm not using Hydrogen for my front-end. It's a custom React app on WordPress

It sounds like passing the `customerAccessToken` to the `buyerIdentity` mutation should work, but doesn't? Not sure where this stands.

**Reply by Judimax** [reply]
Hey I sent you an email I got my setup working and would like to try to help if I can. Im having issues with shopify subscriptions, maybe we have mutual interest maybe I can learn a thing or two

**Comment by JSsantana** [comment]
Hey folks, anything new about this topic? I'm with the same issue. The New Customer API 2024-10 does not keep the customer authenticated when goes to checkout page.

**Comment by JSsantana** [comment]
**MY SOLUTION (workaround):**
For those who are having issues when trying to keep users signed in during a redirect to the checkout page from a headless storefront, here’s a what worked for me: 

The [Shopify Customer API doc](https://shopify.dev/docs/api/customer#authentication) mentions appending `logged_in=true `to the checkoutUrl to maintain user authentication, and while this is correct, it overlooks an important detail. The checkoutUrl returned from the Cart query in the Storefront API (2024-10) is formatted as a `cart/` URL, not a `checkout`/ URL. If you append logged_in=true directly to this cart/ URL, it will not work as intended for authentication. You need to manually transform the cart URL into a checkout URL and then append logged_in=true. When I did this, the customer stayed signed in. Good luck!

**Reply by kivocsa99** [reply]
how did u manage to transform the cart to checkout url ?

**Comment by johncraigcole** [comment]
Earlier this month support was added to ensure a buyer is authenticated at checkout provided the `buyerIdentity.customerAccessToken` is provided on the cart object as noted in this [changelog post](https://shopify.dev/changelog/continuous-cart-authentication).

This should work for both classic customer accounts where a customer access token is created from a customer email and password via [customerAccessTokenCreate](https://shopify.dev/docs/api/storefront/2025-01/mutations/customeraccesstokencreate), or with new customer accounts using an [access token obtained with the customer account api](https://shopify.dev/docs/api/customer#step-obtain-access-token).

**Comment by up1io** [comment]
Unfortunately, neither the cart-to-checkout URL transformation using the logged_in query parameter nor the customerAccessToken method with buyerIdentity upgrade is working. Has anyone found a solution to this issue? Strangely, the API response indicates that everything is functioning correctly. The shopping cart session is associated with my user identity. However, when I navigate to the checkout page, the user is no longer logged in.

**Reply by matt-meagher** [reply]
Same. Sometimes this works when using GraphiQL, but that is not consistent. 
Unless I missed something this follows the documentation outlined in the Changelog
I even set up a route in remix to do this and it has yet to work:
```jsx
import {redirect, json} from '@remix-run/react';

export const loader = async ({request, context}) => {
  const cartID = context.cart.getCartId();
  const customerAccessToken = await context.session.get('customerAccessToken');

  const {cartBuyerIdentityUpdate} = await context.storefront.mutate(
    CART_MUTATION,
    {
      variables: {
        cartID,
        buyerIdentity: {
          customerAccessToken
        }
      }
    }
  )

  const {cart} = await context.storefront.query(
    CHECKOUT_URL_QUERY,
    {
      variables: {
        cartID: cartBuyerIdentityUpdate.cart.id
      }
    }
  )
  return redirect(cart.checkoutUrl);
};

export default function CheckoutUrl() {
  return null;
}

const CHECKOUT_URL_QUERY = `#graphql
  query checkoutURL(
    $cartID: ID!
  ) {
    cart(id: $cartID) {
      checkoutUrl
      id
      buyerIdentity {
        email
        phone
        countryCode
        customer {
          acceptsMarketing
          createdAt
          displayName
          email
          id
        }
      }
    }
  }
`;

const CART_MUTATION = `#graphql
  mutation cartBuyerIdentityUpdate ($cartID: ID!, $buyerIdentity: CartBuyerIdentityInput!) {
    cartBuyerIdentityUpdate(
      cartId: $cartID
      buyerIdentity: $buyerIdentity
    ) {
      cart {
        id
        checkoutUrl
      }
      userErrors {
        field
        message
      }
    }
  }
`
```

**Reply by electricenjindevops** [reply]
```
GPT says....

context.cart.getCartId() is not guaranteed to be defined in a loader
Shopify Hydrogen/Remix context methods like context.cart.getCartId() are client-specific and may not work in a loader, which runs server-side. If no cart is initialized on the server yet, this returns undefined.

This import is incorrect. @remix-run/react is for components. In a loader, you should import from:

import {redirect, json} from '@remix-run/node';


This can cause silent errors or inconsistent behavior, depending on your bundler and environment.


Shopify will not return a cart object if the mutation fails (e.g., invalid/expired customerAccessToken). You’re accessing [cartBuyerIdentityUpdate.cart.id](http://cartbuyeridentityupdate.cart.id/) directly without checking userErrors.


if (cartBuyerIdentityUpdate.userErrors.length > 0) {
  throw new Error(cartBuyerIdentityUpdate.userErrors[0].message);
}

You already have checkoutUrl in cartBuyerIdentityUpdate.cart. You don’t need to re-query the cart.

return redirect(cartBuyerIdentityUpdate.cart.checkoutUrl);



Fixed version (from GPT):

import {redirect, json} from '@remix-run/node';

export const loader = async ({request, context}) => {
  const cartID = context.cart.getCartId();
  const customerAccessToken = await context.session.get('customerAccessToken');

  if (!cartID || !customerAccessToken) {
    throw new Error('Missing cartID or customerAccessToken');
  }

  const {cartBuyerIdentityUpdate} = await context.storefront.mutate(
    CART_MUTATION,
    {
      variables: {
        cartID,
        buyerIdentity: {
          customerAccessToken
        }
      }
    }
  );

  if (cartBuyerIdentityUpdate.userErrors.length > 0) {
    throw new Error(cartBuyerIdentityUpdate.userErrors[0].message);
  }

  const checkoutUrl = cartBuyerIdentityUpdate.cart.checkoutUrl;
  return redirect(checkoutUrl);
};
```

**Comment by electricenjindevops** [comment]
GPT says....

*context.cart.getCartId() is not guaranteed to be defined in a loader*

Shopify Hydrogen/Remix context methods like context.cart.getCartId() *are
client-specific* and may not work in a loader, which runs *server-side*. If
no cart is initialized on the server yet, this returns undefined.

This import is incorrect. @remix-run/react is for components. In a *loader*,
you should import from:

import {redirect, json} from ***@***.***/node';

This can cause silent errors or inconsistent behavior, depending on your
bundler and environment.

Shopify *will not return a cart object if the mutation fails* (e.g.,
invalid/expired customerAccessToken). You’re accessing
cartBuyerIdentityUpdate.cart.id directly without checking userErrors.

if (cartBuyerIdentityUpdate.userErrors.length > 0) {
  throw new Error(cartBuyerIdentityUpdate.userErrors[0].message);
}

You already have checkoutUrl in cartBuyerIdentityUpdate.cart. You don’t
need to re-query the cart.

return redirect(cartBuyerIdentityUpdate.cart.checkoutUrl);


*Fixed version (from GPT):*

import {redirect, json} from ***@***.***/node';

export const loader = async ({request, context}) => {
  const cartID = context.cart.getCartId();
  const customerAccessToken = await
context.session.get('customerAccessToken');

  if (!cartID || !customerAccessToken) {
    throw new Error('Missing cartID or customerAccessToken');
  }

  const {cartBuyerIdentityUpdate} = await context.storefront.mutate(
    CART_MUTATION,
    {
      variables: {
        cartID,
        buyerIdentity: {
          customerAccessToken
        }
      }
    }
  );

  if (cartBuyerIdentityUpdate.userErrors.length > 0) {
    throw new Error(cartBuyerIdentityUpdate.userErrors[0].message);
  }

  const checkoutUrl = cartBuyerIdentityUpdate.cart.checkoutUrl;
  return redirect(checkoutUrl);
};

**Comment by Muneefm** [comment]
@johncraigcole

We have implemented this using the latest APIs. It works when the checkout is opened for the first time, but if I close it and open the checkout again, it redirects to the login page.

We have verified that the cartBuyerIdentity is correct and that the latest checkout URL is being opened.

This issue also occurs when using the Checkout Sheet Kit in React Native.

We are testing this in a React Native environment.

* issue is only present for merchants with guest checkout disabled.

**Comment by MMaxCV** [comment]
This solved the same issue discussed in the issue:
`window.location.href = cart.checkoutUrl?.replace("www", "checkout").concat("&logged_in=true");`
